File: eval/run_eval.js
/**
 * Automated Evaluation Runner for Discord AI Assistant Benchmark
 * Executes 25 Golden Set test cases and evaluates model accuracy.
 */

const fs = require('fs');
const path = require('path');
const { AgentEngine } = require('../fully_functional_webapp/agent_engine.js');

function runEvaluation() {
    console.log("================================================================================");
    console.log("🚀 CHẠY BỘ KIỂM THỬ BENCHMARK GOLDEN SET (25 TEST CASES) FOR TRỢ LÝ DISCORD");
    console.log("================================================================ shower\n");

    const goldenSetPath = path.join(__dirname, 'golden_set.json');
    if (!fs.existsSync(goldenSetPath)) {
        console.error("❌ Không tìm thấy file golden_set.json!");
        process.exit(1);
    }

    const testCases = JSON.parse(fs.readFileSync(goldenSetPath, 'utf-8'));
    const engine = new AgentEngine();

    let passedCount = 0;
    let behaviorMatches = 0;
    let sourceMatches = 0;
    let totalScore = 0;

    const results = [];

    testCases.forEach((tc, idx) => {
        const res = engine.runAgent(tc.question);
        
        const behaviorPass = (res.behavior === tc.expected_behavior);
        let sourcePass = true;

        if (tc.expected_source) {
            sourcePass = (res.citation === tc.expected_source);
        }

        const isPassed = behaviorPass && sourcePass;

        if (isPassed) passedCount++;
        if (behaviorPass) behaviorMatches++;
        if (sourcePass) sourceMatches++;

        const caseScore = isPassed ? 4 : (behaviorPass ? 2 : 0); // 4 pts each for 25 cases = 100 pts total
        totalScore += caseScore;

        results.push({
            id: tc.id,
            category: tc.category,
            question: tc.question,
            expected_behavior: tc.expected_behavior,
            actual_behavior: res.behavior,
            expected_source: tc.expected_source || "N/A",
            actual_source: res.citation || "N/A",
            passed: isPassed,
            answer_preview: res.answer.substring(0, 80) + "..."
        });

        const statusIcon = isPassed ? "✅ PASS" : "❌ FAIL";
        console.log(`[Case ${String(tc.id).padStart(2, '0')}] ${statusIcon} | Category: ${tc.category.padEnd(14)} | Question: "${tc.question}"`);
        if (!isPassed) {
            console.log(`          ↳ Expected Behavior: ${tc.expected_behavior} | Actual: ${res.behavior}`);
            if (tc.expected_source) {
                console.log(`          ↳ Expected Source  : ${tc.expected_source} | Actual: ${res.citation}`);
            }
        }
    });

    const passRate = Math.round((passedCount / testCases.length) * 100);

    console.log("\n================================================================================");
    console.log("📊 KẾT QUẢ ĐO LƯỜNG VÀ ĐIỂM SỐ BENCHMARK THỰC TẾ");
    console.log("================================================================================");
    console.log(` Total Test Cases  : ${testCases.length}`);
    console.log(` Passed Test Cases : ${passedCount} / ${testCases.length}`);
    console.log(` Behavior Accuracy : ${Math.round((behaviorMatches / testCases.length) * 100)}%`);
    console.log(` Citation Accuracy : ${Math.round((sourceMatches / testCases.length) * 100)}%`);
    console.log(` Grounded Pass Rate: ${passRate}%`);
    console.log(` 🏆 TOTAL SCORE    : ${totalScore} / 100 ĐIỂM`);
    console.log("================================================================================\n");

    const summaryReport = {
        timestamp: new Date().toISOString(),
        total_cases: testCases.length,
        passed_cases: passedCount,
        pass_rate_percent: passRate,
        total_score_points: totalScore,
        max_score_points: 100,
        behavior_accuracy_percent: Math.round((behaviorMatches / testCases.length) * 100),
        citation_accuracy_percent: Math.round((sourceMatches / testCases.length) * 100),
        details: results
    };

    const outputPath = path.join(__dirname, 'eval_results.json');
    fs.writeFileSync(outputPath, JSON.stringify(summaryReport, null, 2), 'utf-8');
    console.log(`📁 Báo cáo chi tiết đã được lưu vào: ${outputPath}\n`);

    return summaryReport;
}

if (require.main === module) {
    runEvaluation();
}

module.exports = { runEvaluation };

File: eval/run_eval.py
import json
import os
import sys

# Thêm thư mục gốc vào sys.path để import codebase
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from codebase.rag_engine import RAGEngine

def run_evaluation():
    """
    Script tự động chạy 25 test cases trong eval/golden_set.json
    qua codebase/rag_engine.py và tính % Grounded Rate.
    """
    golden_set_path = "eval/golden_set.json"
    if not os.path.exists(golden_set_path):
        print("Không tìm thấy file eval/golden_set.json")
        return

    with open(golden_set_path, "r", encoding="utf-8") as f:
        test_cases = json.load(f)

    rag = RAGEngine()
    total = len(test_cases)
    high_confidence_count = 0
    low_confidence_count = 0
    results = []

    for case in test_cases:
        query_text = case["question"]
        expected_behavior = case.get("expected_behavior", "answer_with_citation")
        
        res = rag.query(query_text)
        status = res["status"]
        score = res["confidence_score"]
        
        passed = False
        if expected_behavior == "answer_with_citation" and status == "HIGH_CONFIDENCE":
            passed = True
            high_confidence_count += 1
        elif expected_behavior in ["escalate_to_ta", "refuse_politely", "ask_clarification", "rate_limit_warning", "answer_scope", "refuse_sensitive"] and status == "LOW_CONFIDENCE":
            passed = True
            low_confidence_count += 1
        elif status == "HIGH_CONFIDENCE":
            high_confidence_count += 1
        else:
            low_confidence_count += 1

        results.append({
            "id": case["id"],
            "question": query_text,
            "category": case.get("category", "general"),
            "expected_behavior": expected_behavior,
            "actual_status": status,
            "confidence_score": score,
            "citation": res.get("citation", ""),
            "passed": passed
        })

    passed_count = sum(1 for r in results if r["passed"])
    pass_rate = (passed_count / total) * 100 if total > 0 else 0

    eval_report = {
        "total_test_cases": total,
        "passed_cases": passed_count,
        "failed_cases": total - passed_count,
        "pass_rate_percent": round(pass_rate, 2),
        "high_confidence_responses": high_confidence_count,
        "low_confidence_escalations": low_confidence_count,
        "details": results
    }

    # Lưu kết quả vào eval/eval_results.json
    with open("eval/eval_results.json", "w", encoding="utf-8") as f:
        json.dump(eval_report, f, ensure_ascii=False, indent=2)

    # Xuất báo cáo đẹp dạng Markdown vào eval/eval_results.md
    md_content = f"""# 📊 BÁO CÁO KẾT QUẢ KIỂM THỬ EVAL (LƯỢT 1 - BASELINE)

- **Tổng số Test Cases:** {total} cases (từ `eval/golden_set.json`)
- **Số case ĐẠT (Passed):** {passed_count} / {total}
- **Tỷ lệ Đạt (Grounded & Escalation Rate):** **{pass_rate:.1f}%**
- **Luồng High Confidence (Có căn cứ):** {high_confidence_count} cases
- **Luồng Low Confidence (Từ chối & Tag TA):** {low_confidence_count} cases

## Chi Tiết 25 Cases Kiểm Thử:

| ID | Câu Hỏi | Kỳ Vọng | Thực Tế | Confidence Score | Kết Quả |
|---|---|---|---|:---:|:---:|
"""
    for r in results:
        status_icon = "✅ ĐẠT" if r["passed"] else "❌ CHƯA ĐẠT"
        md_content += f"| {r['id']} | `{r['question']}` | {r['expected_behavior']} | {r['actual_status']} | {r['confidence_score']} | {status_icon} |\n"

    with open("eval/eval_results.md", "w", encoding="utf-8") as f:
        f.write(md_content)

    print("EVAL COMPLETE! Pass Rate:", round(pass_rate, 1), "%")

if __name__ == "__main__":
    run_evaluation()
